server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,82 @@
+(ns instant.grouped-queue
+  (:refer-clojure :exclude [peek])
+  (:import
+   (java.util.concurrent LinkedBlockingQueue TimeUnit)
+   (java.util.concurrent.atomic AtomicInteger)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [{:keys [group-fn]}]
+  {:size (AtomicInteger. 0)
+   :group-fn group-fn
+   :group-key->subqueue (atom {})
+   :dispatch-queue (LinkedBlockingQueue.)})
+
+(defn size [{:keys [size] :as _grouped-q}]
+  (.get size))
+
+(defn put! [{:keys [group-fn dispatch-queue group-key->subqueue size]
+             :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (do (.incrementAndGet size)
+          (.put dispatch-queue [:item item]))
+
+      ;; This item will be grouped on `group-key` 
+      (let [_ (.incrementAndGet size)
+            [prev] (swap-vals! group-key->subqueue
```

Comment:
One downside of a single queue atom is that there will be a lot of contention on the atom because every event will have to touch the same atom twice.

Reply:
You might want to add a `locking` around it--I read somewhere that it played more nicely with virtual threads.

Reply:
c.f https://ericnormand.me/guide/clojure-virtual-threads#:~:text=Don%27t%20access%20synchronized%20blocks%20or%20methods

I thought `locking` wasn't good for virtual threads

Reply:
For posterity: 

https://mikemybytes.com/2024/02/28/curiosities-of-java-virtual-threads-pinning-with-synchronized/